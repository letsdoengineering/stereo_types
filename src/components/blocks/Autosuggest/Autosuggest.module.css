/* autosuggest's fixed classNames below here */
.container {
    display: inline-block;
    width: 100%;
}

.input {
    border: none;
    font-size: 1rem;
}

.suggestions-list {
    background: white;
    border: 1px var(--outline-gray) solid;
    margin-top: 1rem;
    max-width: 34rem;
    min-width: 20rem;
    position: absolute;
}

.suggestion {
    background-color: white;
    display: flex;
    padding: 0.35rem 0.5rem;
}

.suggestion--highlighted {
    background-color: var(--outline-gray);
    cursor: pointer;
}

.suggestion:hover {
    background-color: var(--outline-gray);
    cursor: pointer;
}

/* Classnames you can to add styles to the Autosuggest component via this file to Theme prop
/*    container:                'react-autosuggest__container',*/
/*    containerOpen:            'react-autosuggest__container--open',*/
/*    input:                    'react-autosuggest__input',*/
/*    inputOpen:                'react-autosuggest__input--open',*/
/*    inputFocused:             'react-autosuggest__input--focused',*/
/*    suggestionsContainer:     'react-autosuggest__suggestions-container',*/
/*    suggestionsContainerOpen: 'react-autosuggest__suggestions-container--open',*/
/*    suggestionsList:          'react-autosuggest__suggestions-list',*/
/*    suggestion:               'react-autosuggest__suggestion',*/
/*    suggestionFirst:          'react-autosuggest__suggestion--first',*/
/*    suggestionHighlighted:    'react-autosuggest__suggestion--highlighted',*/
/*    sectionContainer:         'react-autosuggest__section-container',*/
/*    sectionContainerFirst:    'react-autosuggest__section-container--first',*/
/*    sectionTitle:             'react-autosuggest__section-title'*/
