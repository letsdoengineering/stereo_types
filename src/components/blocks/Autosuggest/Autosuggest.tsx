import React, { FormEvent } from 'react'
import ReactAutosuggest, {
  BlurEvent,
  ChangeEvent,
  SuggestionsFetchRequestedParams,
  RenderSuggestionsContainerParams,
  AutosuggestProps,
} from 'react-autosuggest'

import * as autoSuggestStyles from './Autosuggest.module.css'

export type AutoSuggestProps = {
  /** callback used to get suggestion value from list */
  getSuggestionValue(dataItem: any): string
  /** callback used get current value of input */
  getInputValue(key: string): string
  /** name attribute */
  name: string
  /** what react-autosuggest will render when suggestions are available */
  renderSuggestion(dataIem: any): React.ReactElement
  /** callback used to track state of the input value */
  setInputValue(name: string, value: string): void
  /** callback used to manage the suggestions in the autosuggest dropdown */
  setSuggestionsData: React.Dispatch<React.SetStateAction<never[]>>
  /** suggestions data used by renderSuggestion in the autosuggest dropdown */
  suggestionsData: any[]
  /** callback used to fetch suggestions data from an external source */
  suggestionsFetchRequest(value: string): void
  /** callback used to render a custom input */
  renderInputComponent: AutosuggestProps<any, any>['renderInputComponent']
}

const Autosuggest = ({
  getSuggestionValue,
  getInputValue,
  name,
  renderSuggestion,
  setInputValue,
  setSuggestionsData,
  suggestionsData,
  suggestionsFetchRequest,
  renderInputComponent,
}: AutoSuggestProps): JSX.Element => {
  const inputProps = {
    id: `${name}-auto-suggest-input`,
    value: getInputValue(name) || '',
    onChange: (event: FormEvent<HTMLElement>, params: ChangeEvent): void =>
      setInputValue(name, params?.newValue || ''),
    onBlur: (event: FormEvent<HTMLElement>, params: BlurEvent<any> | undefined): void => {
      const suggestion = params?.highlightedSuggestion
      if (suggestion) getSuggestionValue(suggestion)
    },
  }

  const onSuggestionsFetchRequested = (request: SuggestionsFetchRequestedParams): void => {
    const { value } = request // Also has variable 'reason'<string> to see what action caused this function call.
    suggestionsFetchRequest(value) // Calls provided function which should take a substring value to be used to fetch suggestion data.
  }

  const shouldRenderSuggestions = (value: string, reason: string): boolean => {
    if (reason === 'input-focused') return false //  CHECK: do we want to show list when when clicking back into input. Or should leave like this where value need to change first, and/or arrow press up/down.
    return !!value // Do not render list if no value present (if they have deleted the last char)
  }

  const onSuggestionsClearRequested = (): void => {
    // Function called to clear the data fetched for suggestions, occurs on blur and maybe other occasions?
    setSuggestionsData([])
  }

  const renderSuggestionsContainer = ({
    containerProps,
    children,
  }: RenderSuggestionsContainerParams): React.ReactElement => (
    // This is a fix to add aria-hidden when list is empty, else it fails axe accessibility tests.
    <div {...containerProps} role={undefined} aria-hidden={suggestionsData.length === 0}>
      {children}
    </div>
  )

  return (
    <ReactAutosuggest
      shouldRenderSuggestions={shouldRenderSuggestions}
      suggestions={suggestionsData}
      onSuggestionsClearRequested={onSuggestionsClearRequested}
      onSuggestionsFetchRequested={onSuggestionsFetchRequested}
      renderSuggestionsContainer={renderSuggestionsContainer}
      highlightFirstSuggestion={true}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      inputProps={inputProps}
      theme={autoSuggestStyles}
      renderInputComponent={renderInputComponent}
    />
  )
}

export default Autosuggest
