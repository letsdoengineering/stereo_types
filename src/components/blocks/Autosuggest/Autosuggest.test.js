import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'

import { axe } from 'jest-axe'
import Autosuggest from './Autosuggest'

let suggestionData = [{ name: 'one' }, { name: 'two' }, { name: 'three' }]
let getInputValue = () => 'a'

const setSuggestionData = () => null
const setInputValue = jest.fn()
const getSuggestionValue = (suggestion) => suggestion.name
const renderSuggestion = (suggestion) => <div>{suggestion.name}</div>
const fetchSuggestions = () => null

const MockCustomInput = React.forwardRef((props, ref) => {
  return (
    <>
      <label htmlFor='destination-auto-suggest-input'>the label</label>
      <input id='destination-auto-suggest-input' {...props} ref={ref} />
    </>
  )
})
MockCustomInput.displayName = 'MockCustomInput'

describe('Autosuggest', () => {
  it('renders an input+label with NO suggestion list initially, when there is no suggestionData', async () => {
    suggestionData = []
    getInputValue = () => null

    render(
      <Autosuggest
        renderInputComponent={(props) => <MockCustomInput {...props} />}
        getSuggestionValue={getSuggestionValue}
        getInputValue={getInputValue}
        name={'destination'}
        renderSuggestion={renderSuggestion}
        setInputValue={setInputValue}
        setSuggestionsData={setSuggestionData}
        suggestionsData={suggestionData}
        suggestionsFetchRequest={fetchSuggestions}
      />
    )

    const destinationField = screen.getByLabelText('the label')
    const suggestionList = screen.queryByRole('listbox')

    expect(destinationField).toBeInTheDocument()
    expect(suggestionList).toBeNull()
  })

  it('renders a suggestions when input has value and suggestions have data', async () => {
    suggestionData = [{ name: 'option one' }, { name: 'option two' }]
    getInputValue = () => 'on'

    render(
      <Autosuggest
        renderInputComponent={(props) => <MockCustomInput {...props} />}
        getSuggestionValue={getSuggestionValue}
        getInputValue={getInputValue}
        name={'destination'}
        renderSuggestion={renderSuggestion}
        setInputValue={setInputValue}
        setSuggestionsData={setSuggestionData}
        suggestionsData={suggestionData}
        suggestionsFetchRequest={fetchSuggestions}
      />
    )

    const destinationField = screen.getByLabelText('the label')
    await userEvent.type(destinationField, 'e', {
      delay: 1,
    })

    expect(setInputValue).toBeCalledWith('destination', 'one')

    const suggestion1 = screen.getByText('option one')
    const suggestion2 = screen.getByText('option two')

    expect(destinationField).toBeInTheDocument()
    expect(suggestion1).toBeInTheDocument()
    expect(suggestion2).toBeInTheDocument()
  })

  it('can click a suggestion to set input value', async () => {
    suggestionData = [{ name: 'option one' }, { name: 'option two' }]
    getInputValue = () => 'on'

    render(
      <Autosuggest
        renderInputComponent={(props) => <MockCustomInput {...props} />}
        getSuggestionValue={getSuggestionValue}
        getInputValue={getInputValue}
        name={'destination'}
        renderSuggestion={renderSuggestion}
        setInputValue={setInputValue}
        setSuggestionsData={setSuggestionData}
        suggestionsData={suggestionData}
        suggestionsFetchRequest={fetchSuggestions}
      />
    )

    const destinationField = screen.getByLabelText('the label')
    await userEvent.type(destinationField, 'e', {
      delay: 1,
    })
    expect(setInputValue).toBeCalledWith('destination', 'one')

    const suggestion1 = screen.getByText('option one')

    await waitFor(() => userEvent.click(suggestion1))
    expect(setInputValue).toBeCalledWith('destination', 'option one')
  })

  it('has not changed the JSX layout of the page from previous snapshot', () => {
    const tree = renderer
      .create(
        <Autosuggest
          renderInputComponent={(props) => <MockCustomInput {...props} />}
          getSuggestionValue={getSuggestionValue}
          getInputValue={getInputValue}
          name={'destination'}
          renderSuggestion={renderSuggestion}
          setInputValue={setInputValue}
          setSuggestionsData={setSuggestionData}
          suggestionsData={suggestionData}
          suggestionsFetchRequest={fetchSuggestions}
        />
      )
      .toJSON()
    // need to get list showing before snapshot
    expect(tree).toMatchSnapshot()
  })

  it('renders the page DOM that is accessible', async () => {
    const { container } = render(
      <Autosuggest
        renderInputComponent={(props) => <MockCustomInput {...props} />}
        getSuggestionValue={getSuggestionValue}
        getInputValue={getInputValue}
        name={'destination'}
        renderSuggestion={renderSuggestion}
        setInputValue={setInputValue}
        setSuggestionsData={setSuggestionData}
        suggestionsData={suggestionData}
        suggestionsFetchRequest={fetchSuggestions}
      />
    )
    const destinationField = screen.getByLabelText('the label')
    userEvent.type(destinationField, 'a')
    /** aria-input-field-name rule ignored as this component on its own does not have a label
     * and should be labelled by another component */
    const results = await axe(container, {
      rules: {
        'aria-input-field-name': { enabled: false },
      },
    })
    expect(results).toHaveNoViolations()
  })
})
