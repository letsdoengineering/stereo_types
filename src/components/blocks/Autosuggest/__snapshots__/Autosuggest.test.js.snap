// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosuggest has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="react-autowhatever-1"
  role="combobox"
  style={Object {}}
>
  <label
    htmlFor="destination-auto-suggest-input"
  >
    the label
  </label>
  <input
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-controls="react-autowhatever-1"
    autoComplete="off"
    id="destination-auto-suggest-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
    type="text"
    value="on"
  />
  <div
    aria-hidden={false}
    id="react-autowhatever-1"
    style={Object {}}
  />
</div>
`;
